src/librustc_mir/transform/type_check.rs

```diff
@@ -0,0 +1,549 @@
+        }
+    }
+
+    fn check_terminator(&self,
+                        mir: &Mir<'tcx>,
+                        term: &Terminator<'tcx>) {
+        debug!("check_terminator: {:?}", term);
+        let tcx = self.tcx();
+        match *term {
+            Terminator::Goto { .. } |
+            Terminator::Resume |
+            Terminator::Return |
+            Terminator::Drop { .. } => {}
+            Terminator::If { ref cond, .. } => {
+                let cond_ty = mir.operand_ty(tcx, cond);
+                match cond_ty.sty {
+                    ty::TyBool => {}
+                    _ => {
+                        span_mirbug!(self, term, "bad If ({:?}, not bool", cond_ty);
+                    }
+                }
+            }
+            Terminator::SwitchInt { ref discr, switch_ty, .. } => {
+                let discr_ty = mir.lvalue_ty(tcx, discr).to_ty(tcx);
+                if let Err(terr) = infer::can_mk_subty(self.infcx, discr_ty, switch_ty) {
+                    span_mirbug!(self, term, "bad SwitchInt ({:?} on {:?}): {:?}",
+                                 switch_ty, discr_ty, terr);
+                }
+            }
+            Terminator::Switch { ref discr, adt_def, .. } => {
+                let discr_ty = mir.lvalue_ty(tcx, discr).to_ty(tcx);
+                match discr_ty.sty {
+                    ty::TyEnum(def, _) if def == adt_def => {},
+                    _ => {
+                        span_mirbug!(self, term, "bad Switch ({:?} on {:?})",
+                                     adt_def, discr_ty);
+                    }
+                }
+            }
+            Terminator::Call { ref func, ref args, ref destination, .. } => {
+                let func_ty = mir.operand_ty(tcx, func);
+                debug!("check_terminator: call, func_ty={:?}", func_ty);
+                let func_ty = match func_ty.sty {
+                    ty::TyBareFn(_, func_ty) => func_ty,
+                    _ => {
+                        span_mirbug!(self, term, "call to non-function {:?}", func_ty);
+                        return;
+                    }
+                };
+                let sig = tcx.erase_late_bound_regions(&func_ty.sig);
+                self.check_call_dest(mir, term, &sig, destination);
+
+                if self.is_box_free(func) {
+                    self.check_box_free_inputs(mir, term, &sig, args);
+                } else {
+                    self.check_call_inputs(mir, term, &sig, args);
+                }
+            }
+        }
+    }
+
+    fn check_call_dest(&self,
+                       mir: &Mir<'tcx>,
+                       term: &Terminator<'tcx>,
+                       sig: &ty::FnSig<'tcx>,
+                       destination: &Option<(Lvalue<'tcx>, BasicBlock)>) {
+        let tcx = self.tcx();
+        match (destination, sig.output) {
+            (&Some(..), ty::FnDiverging) => {
+                span_mirbug!(self, term, "call to diverging function {:?} with dest", sig);
```

Comment:
@arielb1 

> Plus x is not actually initialized, and in fact the assignment is removed by clear_dead_blocks.

Yeah, so I was thinking about this. I am not sure yet it stripping out unreachable code like this is a good strategy long-term -- though I think it MAY be :).

Another alternative would be to have some sort of special rvalue like `unreachable: T` that we can use as the "initializer" for these sorts of cases. (Ideally, though, we'd insert assignments from this initializer in some clever way; the code typeck uses terrifies me.) The reason is that I think we want to type-check that unreachable code nonetheless.

Right now, of course, we're a bit inconsistent, as you pointed out. On the one hand, I think it would be a big breaking change to get strict about type-checking in unreachable scenarios like this:

```rust
let v = if ... { panic!(); } else { 3 }; // note that the "then" branch technically would have type `()`
```

(Though I know that I myself have been confused about this numerous types, since I expect the `;` to cause a type-check error, but it often does not -- though I think I have seen it be required sometimes.)

Moreover, we apparently accept accesses to uninitialized variables in unreachable code, at least sometimes, since [this type-checks](http://is.gd/enorlG):

```rust
fn main() {
    let x: ();
    return;
    use_it(x);
}

fn use_it(x: ()) { }
```

However, this program [DOES yield an error](http://is.gd/AaTXEX):

```rust
fn main() {
    let x: String = format!("foo");
    if false {
        return;
        use_it(x);
    }
    use_it(x);
}

fn use_it(x: String) { }
```

Moreover, as part of the (eventual) move to non-lexical lifetimes, I expect we will be deferring all region checking to be done on the MIR, and hence stripping out unreachable code would definitely affect the set of constraints that get generated. OTOH, it would basically just mean that more things type-check, and in particular it may wind up matching programmer intuition better, since of course the programmer (probably) knows that the code in question is unreachable.